R/normalize.R

```diff
@@ -165,3 +165,80 @@ checkMatrix <- function(x) {
     stop("Matrix data must have exactly two columns", call. = FALSE)
   }
 }
+
+
+# ==== Multi-polygon conversion generic functions ====
+#
+# The return value from the polygonData generic function is a list of
+# multipolygons, plus a bbox attribute.
+#
+# We want to implement polygonData generics for:
+#
+# - lists of multipolygons
+# - individual multipolygons
+# - lists of polygons
+# - individual polygons
+# - lists of multipolylines
+# - individual multipolylines
+# - lists of polylines
+# - individual polylines
+#
+# The previous implementation of this logic tried to directly implement
+# polygonData for each of the above (or at least as many as we could until the
+# scheme fell apart). This doesn't work because the shape of the return value
+# of polygonData must always be the same (a list of multipolygons) and always
+# includes the bbox attribute which is not needed for inner data structures.
+# In other words, polygonData.MULTIPOLYGON can't just do something like
+# lapply(obj, polygonData.POLYGON) because polygonData.POLYGON has too much
+# structure.
+#
+# The new scheme defines a family of conversion functions:
+#
+# - to_multipolygon_list
+# - to_multipolygon
+# - to_polygon
+# - to_ring
+#
+# Each of the specific sp/sf classes need only implement whichever ONE of those
+# actually makes sense (e.g. to_multipolygon_list.sfc,
+# to_multipolygon.MULTIPOLYGON, to_polygon.POLYGON, to_ring.LINESTRING). The
+# higher-level polygonData wrappers will simply call to_multipolygon_list(x),
+# and the default implementations of those methods will fall through to the next
+# level until a match is found.
+
+to_multipolygon_list <- function(x) {
```

Comment:
@tim-salabim,@jcheng5 unexported should be fine.  I don't think we will need these in `mapedit`.  Thanks for all your work on this.  `sf` is definitely the future!